### DISCUSSION TITLE
16 button 4 potentiometer midi controller transposer

### DISCUSSION THREAD

**Discussion by brosekerg** [tttapa/Control-Surface#412] [Q&A] [answered]
Hello,

I need someone to tell me if this is working code for a midi controller with 16 note  buttons, plus 2 extra buttons for incrementing or decrementing the transposer so i have access to more octaves. It also has 4 CCpots for fx. im using the cd74hc4067 mux, so pay particular attention to that part of the code because i need to know if i have that coded right. Am i supposed to sequentially program the buttons in order according to the mux input channels as i have done? or am i supposed to just go from pins 0-24 sequentially to set the notes instead? Thank you!

(Arduino Uno)
```cpp
#include <Control_Surface.h>
USBMIDI_Interface midi;
#include <Arduino_Helpers.h>
using namespace MIDI_Notes;
#include <AH/Hardware/ExtendedInputOutput/AnalogMultiplex.hpp>

CD74HC4067 mux1 = { 6, {2, 3, 4, 5} };

Transposer<-12, +12> transposer;
IncrementDecrementSelector<transposer.getNumberOfBanks()> selector = {
  transposer, {7, 8},
  Wrap::Clamp,
};
 
Bankable::NoteButton buttons[] {
  {transposer, mux1.pin(8), {note(C, 3), CHANNEL_1}},
  {transposer, mux1.pin(7), {note(Db, 3), CHANNEL_1}},
  {transposer, mux1.pin(6), {note(D, 3), CHANNEL_1}},
  {transposer, mux1.pin(5), {note(Eb, 3), CHANNEL_1}},
  {transposer, mux1.pin(4), {note(E, 3), CHANNEL_1}},
  {transposer, mux1.pin(3), {note(F, 3), CHANNEL_1}},
  {transposer, mux1.pin(2), {note(Gb, 3), CHANNEL_1}},
  {transposer, mux1.pin(1), {note(G, 3), CHANNEL_1}},
  {transposer, mux1.pin(22), {note(Ab, 3), CHANNEL_1}},
  {transposer, mux1.pin(21), {note(A, 3), CHANNEL_1}},
  {transposer, mux1.pin(20), {note(Bb, 3), CHANNEL_1}},
  {transposer, mux1.pin(19), {note(B, 3), CHANNEL_1}},
  {transposer, mux1.pin(18), {note(C, 4), CHANNEL_1}},
  {transposer, mux1.pin(17), {note(Db, 4), CHANNEL_1}},
  {transposer, mux1.pin(16), {note(D, 4), CHANNEL_1}},
  {transposer, mux1.pin(15), {note(Eb, 4), CHANNEL_1}},
};

CCPotentiometer potentiometers[] = {
{A0, 0x10},
{A1, 0x11},
{A2, 0x12},
{A3, 0x13},
};

void setup() {
  // put your setup code here, to run once:
 Control_Surface.begin();
 mux1.begin();
 mux1.pinMode(0, INPUT_PULLUP);
}

void loop() {
 Control_Surface.loop();
}
```
-Greg

**Comment by tttapa** [comment, marked as answer]
The pin numbers you're using are not correct. A 4-bit multiplexer has 16 channels, which you can access using `mux1.pin(0)` through `mux1.pin(15)`. If you haven't already, have a look at the [Getting Started](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d7d/md_pages_Getting-Started.html) guide.

The order of the declarations doesn't matter, only which address is associated with each pin.

**Reply by brosekerg** [reply]
ive looked at all of the documentation already i dont see how to do the mux correctly. mux.pin(0) is the input output is it not? so how can i use that for a channel input if that pin goes to arduino? so yr saying u have it to where u just declare 0-15 and u dont worry about the mux datasheet pinout?

**Reply by brosekerg** [reply]
do i need to run the mux through a for loop as well as u did in the getting started? my arduino cant use serial because i flashed hiduino.

**Reply by brosekerg** [reply]
because if we do it according to the datasheet for mux i would be using channels 0-15 correctly right? but once this issue is fixed, does it appear to be working code for my goals?

**Reply by tttapa** [reply]
> ive looked at all of the documentation already i dont see how to do the mux correctly. mux.pin(0) is the input output is it not? so how can i use that for a channel input if that pin goes to arduino? so yr saying u have it to where u just declare 0-15 and u dont worry about the mux datasheet pinout?

`mux.pin(0)` has nothing to do of the pinout of the specific multiplexer package you're using. `pin(0)` is simply the input pin of the first channel of the multiplexer.

> do i need to run the mux through a for loop as well as u did in the getting started? my arduino cant use serial because i flashed hiduino.

I'm not sure what you mean, there aren't any for loops in the getting started guide. You don't have to set any pinModes, this is handled automatically by the `NoteButton` class.  
You don't need any Arduino Helpers headers, and you don't have to call `mux1.begin()` either, this is done automatically by calling `Control_Surface.begin()`:

```cpp
#include <Control_Surface.h>
using namespace MIDI_Notes;

USBMIDI_Interface midi;

CD74HC4067 mux1 = { 6, {2, 3, 4, 5} };

Transposer<-12, +12> transposer;
IncrementDecrementSelector<transposer.getNumberOfBanks()> selector = {
  transposer, {7, 8},
  Wrap::Clamp,
};
 
Bankable::NoteButton buttons[] {
  ...
};

CCPotentiometer potentiometers[] {
  {A0, 0x10},
  {A1, 0x11},
  {A2, 0x12},
  {A3, 0x13},
};

void setup() {
 Control_Surface.begin();
}

void loop() {
 Control_Surface.loop();
}
```

**Reply by brosekerg** [reply]
thank u so much for ur help bro!! and this should work on an Uno right? i look at yr boards and u dont have uno listed. like i said i cant test it til the 4067 gets here in mail, but i wanna be confident it works haha. sorry im a pain in the behind i know lol

**Reply by tttapa** [reply]
It should work on an Uno if you have enough RAM. Control Surface requires quite a bit of memory. The Arduino IDE will tell you how much RAM your sketch uses, Control Surface doesn't use any heap allocation in most cases, so checking the static memory gives a good estimate of the actual memory usage. You will need at least some free RAM for the stack, though.

The Arduino Uno is listed in the [supported boards](https://github.com/tttapa/Control-Surface#supported-boards), if an example supports the AVR architecture, it supports the Arduino Uno.

**Comment by brosekerg** [comment]
Okay thank you and yeah last time I checked it used 76% memory and threw me a warning but I figure if I remove the headers and that other thing u told me to that should free up some space, thank you!!




-------- Original message --------

**Reply by tttapa** [reply]
Unfortunately, removing the headers and the extra code won't save any RAM.

**Comment by brosekerg** [comment]
Your transposer class is great many props to u!! I've been stressing over these midi banks for a month haha

-------- Original message --------

**Comment by brosekerg** [comment]
The warning just says low memory stability issues may occur but its 76% so hopefully its fine



Sent via the Samsung Galaxy S9, an AT&T 5G Evolution capable smartphone



-------- Original message --------
